### PR TITLE
Revamp the service configuration process

### PR DESCRIPTION
**Pull request by ASBishop**
The primary goal of the first patch is to eliminate using crudini to merge disparate sets of configuration snippets into a single cinder.conf file. Instead, we take advantage of oslo's ability to merge snippets from a config-dir (/etc/cinder/cinder.conf.d), and leave the original /etc/cinder/cinder.conf file intact.

The second patch adds support for customServiceConfigSecrets, which provide a mechanism for specifying service configs via Secrets. Instead of specifying sensitive config data directly in the customServiceConfig, a cloud admin can place sensitive data in a Secret, and reference the secret by name in the service's customServiceConfigSecrets.

The third patch add watchers on the Secrets, and updates the service config if the Secret changes.

### CODE REVIEW

File: pkg/cinderapi/volumes.go [hunk 1]
```diff
@@ -24,19 +26,46 @@ func GetVolumes(parentName string, name string, extraVol []cinderv1beta1.CinderE
...
+			VolumeSource: corev1.VolumeSource{
+				Secret: &corev1.SecretVolumeSource{
+					SecretName:  secretName,
+					DefaultMode: &config0640AccessMode,
+				},
+			},
+		}
+
+		volumes = append(volumes, secretVol)
+	}
```
Comment (Akrog) [inline pkg/cinderapi/volumes.go:41]: It should be possible to do this is a common method to reuse in all the services, right?

File: pkg/cinderapi/volumes.go [hunk 2]
```diff
@@ -24,19 +26,46 @@ func GetVolumes(parentName string, name string, extraVol []cinderv1beta1.CinderE
...
+		},
+	}
+
+	for idx, secretName := range secretNames {
+		secretMount := corev1.VolumeMount{
+			Name: secretName,
+			// Each secret needs its own MountPath
+			MountPath: "/var/lib/config-data/secret-" + strconv.Itoa(idx),
+			ReadOnly:  true,
+		}
```
Comment (Akrog) [inline pkg/cinderapi/volumes.go:63]: This also seems common to multiple files

File: pkg/cinderbackup/volumes.go [hunk 3]
```diff
@@ -1,16 +1,28 @@
...
 	var config0640AccessMode int32 = 0640
+	var dirOrCreate = corev1.HostPathDirectoryOrCreate
 
-	backupVolumes := []corev1.Volume{
+	volumes := []corev1.Volume{
+		{
+			Name: "var-lib-cinder",
+			VolumeSource: corev1.VolumeSource{
+				HostPath: &corev1.HostPathVolumeSource{
+					Path: "/var/lib/cinder",
```
Comment (Akrog) [inline pkg/cinderbackup/volumes.go:21]: I know we have this on the cinder-volume service, but why is this necessary in Operator deployed cinder?
Is it for NFS mounts?

File: templates/cinder/bin/init.sh [hunk 4]
```diff
@@ -41,18 +41,26 @@ cp ${DEFAULT_DIR}/* ${MERGED_DIR}
 cp -a ${SVC_CFG} ${SVC_CFG_MERGED}
 
 cp ${DEFAULT_DIR}/cinder.conf ${SVC_CFG_MERGED_DIR}/00-default.conf
+
```
Comment (Akrog) [inline templates/cinder/bin/init.sh:44]: Maybe add the empty lines in the previous commit.

File: controllers/cinderapi_controller.go [hunk 5]
```diff
@@ -199,6 +199,39 @@ func (r *CinderAPIReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
...
+		// get all API CRs
+		apis := &cinderv1beta1.CinderAPIList{}
+		listOpts := []client.ListOption{
+			client.InNamespace(namespace),
+		}
+		if err := r.Client.List(context.Background(), apis, listOpts...); err != nil {
+			r.Log.Error(err, "Unable to retrieve API CRs %v")
+			return nil
+		}
+		for _, cr := range apis.Items {
```
Comment (Akrog) [inline controllers/cinderapi_controller.go:217]: Shouldn't we also be checking for `instance.Spec.Secret` and `instance.Spec.TransportURLSecret`?
Or are you thinking of adding that in another PR?

**Review by Akrog — CHANGES_REQUESTED** [review]

File: templates/common/10-cinder_wsgi.conf [hunk 6]
```diff
@@ -1,26 +0,0 @@
-<VirtualHost *:8776>
```
[the remark this answers is not in the document]
Comment (ASBishop) [reply templates/common/10-cinder_wsgi.conf:1]: There were two copies of this file in the tree, and this one is extraneous. The one that gets installed is from templates/cinder/config/10-cinder_wsgi.conf

File: templates/cinder/bin/init.sh [hunk 7]
```diff
@@ -26,51 +26,54 @@ export DBPASSWORD=${DatabasePassword:?"Please specify a DatabasePassword variabl
...
+mkdir -p ${SVC_CFG_MERGED_DIR}
 
-# expect that the common.sh is in the same dir as the calling script
-SCRIPTPATH="$( cd "$(dirname "$0")" >/dev/null 2>&1 ; pwd -P )"
-. ${SCRIPTPATH}/common.sh --source-only
+cp ${DEFAULT_DIR}/* ${MERGED_DIR}
 
 # Copy default service config from container image as base
 cp -a ${SVC_CFG} ${SVC_CFG_MERGED}
```
[the remark this answers is not in the document]
Comment (ASBishop) [reply templates/cinder/bin/init.sh:41]: Without this line, the resulting /etc/cinder directory in the service containers will have *no* cinder.conf. That may be OK if we assume the file is basically a bunch of comments that document every config setting's default value. Still, I see value in adding it to the containers, if for no other reason than it may help people find options that they want to modify when they're debugging a live problem.

File: templates/cinder/bin/init.sh [hunk 8]
```diff
@@ -26,51 +26,54 @@ export DBPASSWORD=${DatabasePassword:?"Please specify a DatabasePassword variabl
 export PASSWORD=${CinderPassword:?"Please specify a CinderPassword variable."}
 export TRANSPORTURL=${TransportURL:-""}
 
-export CUSTOMCONF=${CustomConf:-""}
-
+DEFAULT_DIR=/var/lib/config-data/default
+CUSTOM_DIR=/var/lib/config-data/custom
+MERGED_DIR=/var/lib/config-data/merged
 SVC_CFG=/etc/cinder/cinder.conf
```
[the remark this answers is not in the document]
Comment (ASBishop) [reply templates/cinder/bin/init.sh:32]: Not if we retain L41. I suppose it could be eliminated and just use the "/etc/cinder/cinder.conf" string in L41. The shell variable was referenced a lot more before this patch, so having a variable made sense. Given my vote to retain L41, what are your thoughts on retaining the variable?

File: templates/cinder/bin/init.sh [hunk 9]
```diff
@@ -26,51 +26,54 @@ export DBPASSWORD=${DatabasePassword:?"Please specify a DatabasePassword variabl
...
-SVC_CFG_MERGED=/var/lib/config-data/merged/cinder.conf
+SVC_CFG_MERGED=${MERGED_DIR}/cinder.conf
```
[the remark this answers is not in the document]
Comment (ASBishop) [reply templates/cinder/bin/init.sh:33]: See prior comments.

File: templates/cinder/bin/init.sh [hunk 10]
```diff
@@ -26,51 +26,54 @@ export DBPASSWORD=${DatabasePassword:?"Please specify a DatabasePassword variabl
...
+transport_url = ${TRANSPORTURL}
+
+EOF
 fi
-crudini --set ${SVC_CFG_MERGED} database connection mysql+pymysql://${DBUSER}:${DBPASSWORD}@${DBHOST}/${DB}
-crudini --set ${SVC_CFG_MERGED} keystone_authtoken password $PASSWORD
-crudini --set ${SVC_CFG_MERGED} nova password $PASSWORD
+
 # TODO: service token
 #crudini --set ${SVC_CFG_MERGED} service_user password $CinderPassword
```
[the remark this answers is not in the document]
Comment (ASBishop) [reply templates/cinder/bin/init.sh:71]: Ack, I originally removed it but it snuck back in during a rebase.

File: templates/cinder/bin/init.sh (hunk 4 above)
Comment (ASBishop) [reply templates/cinder/bin/init.sh:44]: Ack.

File: pkg/cinderapi/volumes.go (hunk 1 above)
Comment (ASBishop) [reply pkg/cinderapi/volumes.go:41]: Yeah, I think so! I'll work on that.

File: templates/cinder/bin/init.sh [hunk 11]
```diff
@@ -26,51 +26,54 @@ export DBPASSWORD=${DatabasePassword:?"Please specify a DatabasePassword variabl
...
-crudini --merge ${SVC_CFG_MERGED} < /var/lib/config-data/default/custom.conf
-
-# TODO: a cleaner way to handle this?
-# There might be service-specific extra custom conf that needs to be merged
-# with the main cinder.conf for this particular service
-if [ -n "$CUSTOMCONF" ]; then
-    echo merging /var/lib/config-data/custom/${CUSTOMCONF} into ${SVC_CFG_MERGED}
-    crudini --merge ${SVC_CFG_MERGED} < /var/lib/config-data/custom/${CUSTOMCONF}
+cp ${DEFAULT_DIR}/cinder.conf ${SVC_CFG_MERGED_DIR}/00-default.conf
+# 01-secrets.conf is created below
```
[the remark this answers is not in the document]
Comment (ASBishop) [reply templates/cinder/bin/init.sh]: I'll update the patch to generate the files in numeric order.

File: api/bases/cinder.openstack.org_cinderapis.yaml [hunk 12]
```diff
@@ -56,9 +56,17 @@ spec:
                 description: CustomServiceConfig - customize the service config using
                   this parameter to change service defaults, or overwrite rendered
                   information using raw OpenStack config format. The content gets
-                  added to to /etc/<service>/<service>.conf.d directory as custom.conf
-                  file.
+                  added to to /etc/<service>/<service>.conf.d directory as a custom
+                  config file.
                 type: string
+              customServiceConfigSecrets:
+                description: CustomServiceConfigSecrets - customize the service config
```
[the remark this answers is not in the document]
Comment (ASBishop) [reply api/bases/cinder.openstack.org_cinderapis.yaml]: Yep.

File: pkg/cinderapi/volumes.go (hunk 1 above)
Comment (ASBishop) [reply pkg/cinderapi/volumes.go]: Deduped this in latest patchset

File: pkg/cinderapi/volumes.go (hunk 2 above)
Comment (ASBishop) [reply pkg/cinderapi/volumes.go]: Deduped in latest patchset

File: controllers/cinderapi_controller.go (hunk 5 above)
Comment (Akrog) [reply controllers/cinderapi_controller.go:217]: Maybe we should add a comment mentioning how the other secrets are watched by the cinder-operator.

File: templates/cinder/bin/init.sh (hunk 7 above)
[the remark this answers is not in the document]
Comment (Akrog) [reply templates/cinder/bin/init.sh:41]: I agree that has value, moreover now that we are transparent with the config options in the installer.
Maybe we should rename it to something like `cinder.conf.sample` to indicate the intent of the file and avoid misunderstandings.

File: templates/cinder/bin/init.sh (hunk 8 above)
[the remark this answers is not in the document]
Comment (Akrog) [reply templates/cinder/bin/init.sh:32]: I'm ok if we change it to:
```
 SVC_SAMPLE_CFG=/etc/cinder/cinder.conf.sample
```

File: templates/common/10-cinder_wsgi.conf (hunk 6 above)
[the remark this answers is not in the document]
Comment (Akrog) [reply templates/common/10-cinder_wsgi.conf:1]: Oh, I missed that this was a different file!

File: pkg/cinderbackup/volumes.go (hunk 3 above)
Comment (Akrog) [reply pkg/cinderbackup/volumes.go:19]: Usually I wouldn't want to add this change in this patch, but since I suspect we need it for NFS when volume and backup are running in the same host we can leave it to be the same as TripleO and then evaluate if we can remove it.

File: templates/cinder/bin/init.sh (hunk 8 above)
[the remark this answers is not in the document]
Comment (ASBishop) [reply templates/cinder/bin/init.sh:32]: Sure, that sounds good. In addition, we could create a 1-line cinder.conf file that's just a comment directing people to the cinder.conf.d directory. That might be useful because RHOSP hasn't traditionally used that directory, and people may wonder what happened to the cinder.conf file they're used to seeing.

Thoughts?

File: controllers/cinderapi_controller.go (hunk 5 above)
Comment (ASBishop) [reply controllers/cinderapi_controller.go:217]: Ack.

File: templates/cinder/bin/init.sh [hunk 13]
```diff
@@ -26,51 +26,59 @@ export DBPASSWORD=${DatabasePassword:?"Please specify a DatabasePassword variabl
...
-# TODO: a cleaner way to handle this?
-# Merge custom.conf with cinder.conf, since the Kolla config doesn't seem
-# to allow us to customize the cinder command (it calls httpd instead).
-# Can we just put custom.conf in something like /etc/cinder/cinder.conf.d/custom.conf
-# and have it automatically detected, or would we have to somehow change the call
-# to the cinder binary to tell it to use that custom conf dir?
-echo merging /var/lib/config-data/default/custom.conf into ${SVC_CFG_MERGED}
-crudini --merge ${SVC_CFG_MERGED} < /var/lib/config-data/default/custom.conf
+# Save the default service config from container image as cinder.conf.sample,
+# and direct create a small cinder.conf file that directs people to files
```
Comment (ASBishop) [inline templates/cinder/bin/init.sh:41]: That should be "and create" (not "and direct create"). I can fix this if other updates are necessary, or it could be treated as an awkwardly worded nit.

File: templates/cinder/bin/init.sh [hunk 14]
```diff
@@ -26,51 +26,54 @@ export DBPASSWORD=${DatabasePassword:?"Please specify a DatabasePassword variabl
...
+# 01-secrets.conf is created below
+if [ -f ${DEFAULT_DIR}/custom.conf ]; then
+    cp ${DEFAULT_DIR}/custom.conf ${SVC_CFG_MERGED_DIR}/02-global.conf
+fi
+if [ -f ${CUSTOM_DIR}/custom.conf ]; then
+    cp ${CUSTOM_DIR}/custom.conf ${SVC_CFG_MERGED_DIR}/03-service.conf
 fi
 
 # set secrets
+SVC_CFG_SECRETS=${SVC_CFG_MERGED_DIR}/01-secrets.conf
```
[the remark this answers is not in the document]
Comment (ASBishop) [reply templates/cinder/bin/init.sh]: AFAIK, the init container starts with an empty "merged" directory.

File: templates/cinder/bin/init.sh [hunk 15]
```diff
@@ -48,17 +48,17 @@ EOF
 cp ${DEFAULT_DIR}/cinder.conf ${SVC_CFG_MERGED_DIR}/00-default.conf
 
 # Generate 01-secrets.conf
```
Comment (ASBishop) [inline templates/cinder/bin/init.sh:50]: file name in comment needs to be updated.

File: templates/cinder/bin/init.sh (hunk 8 above)
[the remark this answers is not in the document]
Comment (Akrog) [reply templates/cinder/bin/init.sh:32]: I think it's a great idea.

File: controllers/cinderapi_controller.go [hunk 16]
```diff
@@ -709,6 +726,38 @@ func (r *CinderAPIReconciler) reconcileUpgrade(ctx context.Context, instance *ci
...
+		instance.Status.Conditions.Set(condition.FalseCondition(
+			condition.InputReadyCondition,
+			condition.ErrorReason,
+			condition.SeverityWarning,
+			condition.InputReadyErrorMessage,
+			err.Error()))
+		return ctrl.Result{}, err
+	}
+
+	(*envVars)[secret.Name] = env.SetValue(hash)
```
[the remark this answers is not in the document]
Comment (ASBishop) [reply controllers/cinderapi_controller.go]: Done (and I had to do it in all controllers).

**Review by Akrog — APPROVED** [review]
/lgtm
